From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Fri, 10 Feb 2023 17:45:59 +0200
Subject: [PATCH] Kaiiju Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 57d4789c1eb00cb376adce31ece2b6672c4a12a9..aba8170a8254558b33a5ac6635c30c5cc7e39296 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2030,6 +2030,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Kaiiju start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKaiijuConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kaiiju end
+
         /**
          * Sends the component to the player
          *
