From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Fri, 10 Feb 2023 17:45:59 +0200
Subject: [PATCH] Kaiiju Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1ef4cf3b19d58565e052829358eb5902f070ae02..490a70920dca1cfde4796cc2fb69fdb98942bbdc 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2007,6 +2007,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Kaiiju start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKaiijuConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kaiiju end
+
         /**
          * Sends the component to the player
          *
