From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Fri, 10 Feb 2023 17:45:59 +0200
Subject: [PATCH] Kaiiju Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index f77553fc70bc1f3f18af50e53c0041bbaaebab9a..3607ae96bed9e54b919456ca934cddd5664a5006 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2081,6 +2081,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Kaiiju start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getKaiijuConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Kaiiju end
+
         /**
          * Sends the component to the player
          *
