From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kugge <sofiane.djerbi38@gmail.com>
Date: Sun, 15 Jan 2023 21:42:02 +0100
Subject: [PATCH] Kaiiju AFK Distance


diff --git a/src/main/java/dev/kugge/kaiiju/KaiijuConfig.java b/src/main/java/dev/kugge/kaiiju/KaiijuConfig.java
index 45df8d6569e0d3ca5ceec6cd4de59caeb967e69c..ef36cf46a1b26f9346fac87e1ae115b0c9dcd94f 100644
--- a/src/main/java/dev/kugge/kaiiju/KaiijuConfig.java
+++ b/src/main/java/dev/kugge/kaiiju/KaiijuConfig.java
@@ -133,4 +133,25 @@ public class KaiijuConfig {
         return config.getStringList(key);
     }
 
-}
\ No newline at end of file
+    public static boolean useIdleDistances;
+    private static void getUseIdleDistances() {
+        useIdleDistances = getBoolean("distances.use-afk-distances", false,
+                                  "Whether or not AFK distances should be enabled.",
+                                  "Require \"kick-if-idle: false\" in Purpur config.");
+    }
+
+    public static int idleViewDistance;
+    private static void getIdleViewDistance() {
+        idleViewDistance = getInt("distances.afk-view-distance", 8,
+                                  "View distance for AFK players.",
+                                  "Require \"kick-if-idle: false\" in Purpur config.");
+    }
+
+    public static int idleSimulationDistance;
+    private static void getIdleSimulationDistance() {
+        idleSimulationDistance = getInt("distances.afk-simulation-distance", 8,
+                                        "Simulation distance for AFK players.",
+                                        "Require \"kick-if-idle: false\" in Purpur config.");
+    }
+
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index d62bf95fb067c2e2a4fc9b2757c636bd4ee75bea..abb81d8ce7b6df8c4a8cd1b7add00cb7910ca5eb 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2084,6 +2084,12 @@ public class ServerPlayer extends Player {
     public void resetLastActionTime() {
         this.lastActionTime = Util.getMillis();
         this.setAfk(false); // Purpur
+        // Kaiiju start
+        if (dev.kugge.kaiiju.KaiijuConfig.useIdleDistances) {
+            this.getBukkitEntity().setViewDistance(this.getBukkitEntity().getWorld().getViewDistance());
+            this.getBukkitEntity().setSimulationDistance(this.getBukkitEntity().getWorld().getSimulationDistance());
+        }
+        // Kaiiju end
     }
 
     // Purpur Start
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a8e205af936efe0909c94575e3b7f8ca50cead2f..6d022a878287d3359145fcbb8c1e6f8743393b98 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -461,6 +461,12 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
             // Purpur start
             this.player.setAfk(true);
             if (!this.player.level.purpurConfig.idleTimeoutKick || kickPermissionCache.getUnchecked(this.player.getBukkitEntity())) {
+                // Kaiiju start
+                if (dev.kugge.kaiiju.KaiijuConfig.useIdleDistances) {
+                    this.getCraftPlayer().setViewDistance(dev.kugge.kaiiju.KaiijuConfig.idleViewDistance);
+                    this.getCraftPlayer().setSimulationDistance(dev.kugge.kaiiju.KaiijuConfig.idleSimulationDistance);
+                }
+                // Kaiiju end
                 return;
             }
             // Purpur end
