From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Sat, 11 Feb 2023 02:10:20 +0200
Subject: [PATCH] Kaiiju Lobotomize Configuration


diff --git a/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java b/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java
index e77b6d86f651dc93f5c9dc28eda1e00a6af7ac8d..6e17f49999da4cd6d690b38e3541223050d126b2 100644
--- a/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java
+++ b/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java
@@ -141,4 +141,7 @@ public class KaiijuWorldConfig {
             regionFormatLinearCompressionLevel = 1;
         }
     }
+
+    private void lobotomizeSettings() {
+    }
 }
\ No newline at end of file
