From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Thu, 16 Feb 2023 01:38:59 +0200
Subject: [PATCH] Network configuration


diff --git a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
index b0ee4b9dd8fb40761c21ad43eeb1165cbde84962..22af24dd08a368e2b307840b0330be34f3fe84af 100644
--- a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
+++ b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
@@ -198,4 +198,7 @@ public class KaiijuConfig {
     private static void regionFormatSettings() {
         linearFlushFrequency = getInt("region-format.linear.flush-frequency", linearFlushFrequency);
     }
+
+    private static void networkSettings() {
+    }
 }
