From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sat, 8 Apr 2023 18:10:14 +0200
Subject: [PATCH] Add option to drop contents of shulker items when destroyed.


diff --git a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
index edaeba20d871925ae05f7f1cabb3ac8eb71b4655..26442d7cac765f938a084aa69dd2aadb4a5dd8a3 100644
--- a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
+++ b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
@@ -206,4 +206,10 @@ public class KaiijuConfig {
         sendNullEntityPackets = getBoolean("network.send-null-entity-packets", sendNullEntityPackets);
         alternateKeepAlive = getBoolean("network.alternate-keepalive", sendNullEntityPackets);
     }
+
+    public static boolean shulkerBoxDropContentsWhenDestroyed = true;
+
+    private static void gameplaySettings() {
+        shulkerBoxDropContentsWhenDestroyed = getBoolean("gameplay.shulker-box-drop-contents-when-destroyed", shulkerBoxDropContentsWhenDestroyed);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/item/BlockItem.java b/src/main/java/net/minecraft/world/item/BlockItem.java
index b0204af850ee182773ad458208cccd946ad148d5..b09c8dc9de9baad7c0c18babc2c83ad1ca8f9b9c 100644
--- a/src/main/java/net/minecraft/world/item/BlockItem.java
+++ b/src/main/java/net/minecraft/world/item/BlockItem.java
@@ -288,7 +288,7 @@ public class BlockItem extends Item {
 
     @Override
     public void onDestroyed(ItemEntity entity) {
-        if (this.block instanceof ShulkerBoxBlock) {
+        if (this.block instanceof ShulkerBoxBlock && dev.kaiijumc.kaiiju.KaiijuConfig.shulkerBoxDropContentsWhenDestroyed) {
             ItemStack itemstack = entity.getItem();
             CompoundTag nbttagcompound = BlockItem.getBlockEntityData(itemstack);
 
