From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Thu, 16 Feb 2023 21:28:49 +0200
Subject: [PATCH] Kaiiju Lithium Configuration


diff --git a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
index c213726efa4dd95e6e731e9c7e75fb23c3c4763f..e1fc127ac3fb47f3a4ad84d4b3f3ec25f40389f8 100644
--- a/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
+++ b/src/main/java/dev/kaiijumc/kaiiju/KaiijuConfig.java
@@ -215,4 +215,9 @@ public class KaiijuConfig {
     private static void networkSettings() {
         sendNullEntityPackets = getBoolean("network.send-null-entity-packets", sendNullEntityPackets);
     }
+
+    public static boolean lithiumEnable = true;
+    private static void lithiumSettings() {
+        lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
+    }
 }
\ No newline at end of file
