From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Sun, 7 May 2023 20:04:06 +0300
Subject: [PATCH] Toggle break redstone on top of trap doors early

That patch break vanilla mechanics such as portal slicing.

diff --git a/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java b/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java
index 601b9867d839f4928ba993ee8f00df9c330a0fe3..e7ee85b09aa4438868e068551280b13e84cdfdb8 100644
--- a/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java
+++ b/src/main/java/dev/kaiijumc/kaiiju/KaiijuWorldConfig.java
@@ -157,8 +157,10 @@ public class KaiijuWorldConfig {
     }
 
     public boolean fixVoidTrading = true;
+    public boolean breakRedstoneOnTopOfTrapDoorsEarly = true;
 
     private void gameplaySettings() {
         fixVoidTrading = getBoolean("gameplay.fix-void-trading", fixVoidTrading);
+        breakRedstoneOnTopOfTrapDoorsEarly = getBoolean("gameplay.break-redstone-on-top-of-trap-doors-early", breakRedstoneOnTopOfTrapDoorsEarly);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java b/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
index b9e3c9cd709d42bced85436720ccc7b9c7e49552..5e6d64408a5da3f337d32bb8ac5ac2532b04490c 100644
--- a/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -125,6 +125,7 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                 }
                 // CraftBukkit end
                 boolean open = (Boolean) state.getValue(TrapDoorBlock.OPEN) != flag1; // Paper - break redstone on trapdoors early
+                if (world.kaiijuConfig.breakRedstoneOnTopOfTrapDoorsEarly) { // Kaiiju - trigger break redstone on trapdoors early
                 // Paper start - break redstone on trapdoors early
                 // note: this must run before any state for this block/its neighborus are written to the world
                 // we allow the redstone event to fire so that plugins can block
@@ -143,6 +144,7 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                     }
                 }
                 // Paper end - break redstone on trapdoors early
+                } // Kaiiju  - trigger break redstone on trapdoors early
                 if (open) { // Paper - break redstone on trapdoors early
                     state = (BlockState) state.setValue(TrapDoorBlock.OPEN, flag1);
                     this.playSound((Player) null, world, pos, flag1);
