From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Sofiane H. Djerbi" <46628754+kugge@users.noreply.github.com>
Date: Fri, 19 May 2023 03:38:03 +0300
Subject: [PATCH] Vanilla end portal teleportation


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 26c776826b0b95f54361c6f00062543d92ef2d3d..b03e09fd3f794ab5d5569cea8e376915fe436816 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -411,6 +411,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean collidingWithWorldBorder; // Paper
+    public Vec3 previousVelocity = Vec3.ZERO; // Kaiiju - Vanilla end Teleportation
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -3913,10 +3914,15 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                         (chunks) -> {
                             ServerLevel.makeObsidianPlatform(destination, null, targetPos);
 
+                            // Kaiiju start - Vanilla end teleportation
+                            Vec3 finalPos;
+                            if (this instanceof Player) finalPos = Vec3.atBottomCenterOf(targetPos.below());
+                            else finalPos = Vec3.atBottomCenterOf(targetPos);
+                            // Kaiiju end
                             // the portal obsidian is placed at targetPos.y - 2, so if we want to place the entity
                             // on the obsidian, we need to spawn at targetPos.y - 1
                             portalInfoCompletable.complete(
-                                new PortalInfo(Vec3.atBottomCenterOf(targetPos.below()), Vec3.ZERO, 90.0f, 0.0f, destination, null)
+                                new PortalInfo(finalPos, this.getDeltaMovement(), 90.0f, 0.0f, destination, null) // Kaiiju - Vanilla end teleportation
                             );
                         }
                     );
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index c8d125955754e27da54d95fb5b1cea39ca54b618..3d9555876fb4b89b24719a5f17e2194ecc63eb14 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -47,7 +47,7 @@ import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
-import org.slf4j.Logger;
+import org.bukkit.Bukkit;import org.slf4j.Logger;
 
 import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
@@ -135,6 +135,7 @@ public class FallingBlockEntity extends Entity {
             return;
         }
         // Paper end - fix sand duping
+        Bukkit.getLogger().info("velocity BEFORE:" + this.getDeltaMovement());
         if (this.blockState.isAir()) {
             this.discard();
         } else {
@@ -258,6 +259,7 @@ public class FallingBlockEntity extends Entity {
 
             this.setDeltaMovement(this.getDeltaMovement().scale(0.98D));
         }
+        Bukkit.getLogger().info("velocity AFTER:" + this.getDeltaMovement());
     }
 
     public void callOnBrokenAfterFall(Block block, BlockPos pos) {
