From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Xymb <xymb@endcrystal.me>
Date: Sat, 17 Jun 2023 21:03:04 +0300
Subject: [PATCH] Use Math.floor instead of fastfloor


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index cadb91c7f5ef80aac8513f246fdf481947fa0a17..c1a2d67cb62af94c8bb4dff97f4fcb47542562dd 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -168,13 +168,19 @@ public final class MCUtil {
     }
 
     public static int fastFloor(double x) {
-        int truncated = (int)x;
-        return x < (double)truncated ? truncated - 1 : truncated;
+        // Kaiiju start - Use Math.floor instead of fastfloor
+        return (int)Math.floor(x); // Kaiiju
+        //int truncated = (int)x;
+        //return x < (double)truncated ? truncated - 1 : truncated;
+        // Kaiiju end
     }
 
     public static int fastFloor(float x) {
-        int truncated = (int)x;
-        return x < (double)truncated ? truncated - 1 : truncated;
+        // Kaiiju start - Use Math.floor instead of fastfloor
+        return (int)Math.floor(x); // Kaiiju
+        //int truncated = (int)x;
+        //return x < (double)truncated ? truncated - 1 : truncated;
+        // Kaiiju end
     }
 
     public static float normalizeYaw(float f) {
